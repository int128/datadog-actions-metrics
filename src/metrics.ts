import { Series } from '@datadog/datadog-api-client/dist/packages/datadog-api-client-v1/models/Series'
import { WorkflowRunEvent } from '@octokit/webhooks-definitions/schema'
import { ListJobsForWorkflowRun } from './types'

export const computeWorkflowRunMetrics = (e: WorkflowRunEvent): Series[] => {
  const updatedAt = new Date(e.workflow_run.updated_at).getTime() / 1000
  const tags = [
    `repository_owner:${e.workflow_run.repository.owner.login}`,
    `repository_name:${e.workflow_run.repository.name}`,
    `workflow_name:${e.workflow_run.name}`,
    `event:${e.workflow_run.event}`,
    `conclusion:${e.workflow_run.conclusion}`,
    `branch:${e.workflow_run.head_branch}`,
    `default_branch:${e.workflow_run.head_branch === e.repository.default_branch}`,
  ]
  return [
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.workflow_run.total',
      type: 'count',
      points: [[updatedAt, 1]],
    },
    {
      host: 'github.com',
      tags,
      metric: `github.actions.workflow_run.conclusion.${e.workflow_run.conclusion}_total`,
      type: 'count',
      points: [[updatedAt, 1]],
    },
  ]
}

export const computeJobMetrics = (e: WorkflowRunEvent, listJobsForWorkflowRun: ListJobsForWorkflowRun): Series[] => {
  const series: Series[] = []
  for (const j of listJobsForWorkflowRun.jobs) {
    if (j.completed_at === null) {
      continue
    }

    const completedAt = new Date(j.completed_at).getTime() / 1000
    const tags = [
      `repository_owner:${e.workflow_run.repository.owner.login}`,
      `repository_name:${e.workflow_run.repository.name}`,
      `workflow_name:${e.workflow_run.name}`,
      `event:${e.workflow_run.event}`,
      `branch:${e.workflow_run.head_branch}`,
      `default_branch:${e.workflow_run.head_branch === e.repository.default_branch}`,
      `job_name:${j.name}`,
      `conclusion:${j.conclusion}`,
      `status:${j.status}`,
    ]
    series.push(
      {
        host: 'github.com',
        tags,
        metric: 'github.actions.job.total',
        type: 'count',
        points: [[completedAt, 1]],
      },
      {
        host: 'github.com',
        tags,
        metric: `github.actions.job.conclusion.${j.conclusion}_total`,
        type: 'count',
        points: [[completedAt, 1]],
      }
    )

    const startedAt = new Date(j.started_at).getTime() / 1000
    const duration = completedAt - startedAt
    series.push({
      host: 'github.com',
      tags,
      metric: 'github.actions.job.duration_second',
      type: 'gauge',
      points: [[completedAt, duration]],
    })

    const createdAt = new Date(e.workflow_run.created_at).getTime() / 1000
    const queued = startedAt - createdAt
    series.push({
      host: 'github.com',
      tags,
      metric: 'github.actions.job.queued_duration_second',
      type: 'gauge',
      points: [[completedAt, queued]],
    })
  }
  return series
}

export const computeStepMetrics = (e: WorkflowRunEvent, listJobsForWorkflowRun: ListJobsForWorkflowRun): Series[] => {
  const series: Series[] = []
  for (const job of listJobsForWorkflowRun.jobs) {
    if (job.completed_at === null || job.steps === undefined) {
      continue
    }
    for (const s of job.steps) {
      if (s.started_at == null || s.completed_at == null) {
        continue
      }

      const completedAt = new Date(s.completed_at).getTime() / 1000
      const tags = [
        `repository_owner:${e.workflow_run.repository.owner.login}`,
        `repository_name:${e.workflow_run.repository.name}`,
        `workflow_name:${e.workflow_run.name}`,
        `event:${e.workflow_run.event}`,
        `branch:${e.workflow_run.head_branch}`,
        `default_branch:${e.workflow_run.head_branch === e.repository.default_branch}`,
        `job_name:${job.name}`,
        `step_name:${s.name}`,
        `conclusion:${s.conclusion}`,
        `status:${s.status}`,
      ]
      series.push(
        {
          host: 'github.com',
          tags,
          metric: 'github.actions.step.total',
          type: 'count',
          points: [[completedAt, 1]],
        },
        {
          host: 'github.com',
          tags,
          metric: `github.actions.step.conclusion.${s.conclusion}_total`,
          type: 'count',
          points: [[completedAt, 1]],
        }
      )

      const startedAt = new Date(s.started_at).getTime() / 1000
      const duration = completedAt - startedAt
      series.push({
        host: 'github.com',
        tags,
        metric: 'github.actions.step.duration_second',
        type: 'gauge',
        points: [[completedAt, duration]],
      })
    }
  }
  return series
}
