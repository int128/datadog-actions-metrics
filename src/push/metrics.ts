import { v1 } from '@datadog/datadog-api-client'
import { PushEvent } from '@octokit/webhooks-types'

export const computePushMetrics = (e: PushEvent, now: Date): v1.Series[] => {
  const tags = [
    `repository_owner:${e.repository.owner.login}`,
    `repository_name:${e.repository.name}`,
    `sender:${e.sender.login}`,
    `sender_type:${e.sender.type}`,
    `ref:${e.ref}`,
    `created:${String(e.created)}`,
    `deleted:${String(e.deleted)}`,
    `forced:${String(e.forced)}`,
    `default_branch:${(e.ref === `refs/heads/${e.repository.default_branch}`).toString()}`,
  ]
  const t = now.getTime() / 1000
  return [
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.push.total',
      type: 'count',
      points: [[t, 1]],
    },
  ]
}
