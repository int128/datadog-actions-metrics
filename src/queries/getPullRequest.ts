import assert from 'assert'
import { GetPullRequestQuery, GetPullRequestQueryVariables } from '../generated/graphql.js'
import { Octokit } from '../types.js'

const query = /* GraphQL */ `
  query getPullRequest($owner: String!, $name: String!, $number: Int!) {
    repository(owner: $owner, name: $name) {
      pullRequest(number: $number) {
        commits(first: 1) {
          nodes {
            commit {
              authoredDate
              committedDate
            }
          }
        }
      }
    }
  }
`

export type PullRequestFirstCommit = {
  authoredDate: string
  committedDate: string
}

export const getPullRequestFirstCommit = async (
  o: Octokit,
  v: GetPullRequestQueryVariables,
): Promise<PullRequestFirstCommit | undefined> => {
  const r = await o.graphql<GetPullRequestQuery>(query, v)
  assert(r.repository != null)
  assert(r.repository.pullRequest != null)
  assert(r.repository.pullRequest.commits != null)
  assert(r.repository.pullRequest.commits.nodes != null)
  if (r.repository.pullRequest.commits.nodes.length === 0) {
    return // pull request has no commit
  }
  assert(r.repository.pullRequest.commits.nodes[0] != null)
  return r.repository.pullRequest.commits.nodes[0].commit
}
