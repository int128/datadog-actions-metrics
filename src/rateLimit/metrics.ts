import { v1 } from '@datadog/datadog-api-client'
import { RateLimitResponse } from '../types.js'

type Context = {
  repo: {
    owner: string
    repo: string
  }
}

export const computeRateLimitMetrics = (e: Context, r: RateLimitResponse): v1.Series[] => {
  const t = unixTime(r.headers.date) ?? Math.floor(Date.now() / 1000)
  const tags = [`repository_owner:${e.repo.owner}`, `repository_name:${e.repo.repo}`]

  const series: v1.Series[] = [
    {
      host: 'github.com',
      tags: [...tags, 'resource:core'],
      metric: 'github.actions.api_rate_limit.remaining',
      type: 'gauge',
      points: [[t, r.data.resources.core.remaining]],
    },
    {
      host: 'github.com',
      tags: [...tags, 'resource:core'],
      metric: `github.actions.api_rate_limit.limit`,
      type: 'gauge',
      points: [[t, r.data.resources.core.limit]],
    },
    {
      host: 'github.com',
      tags: [...tags, 'resource:search'],
      metric: 'github.actions.api_rate_limit.remaining',
      type: 'gauge',
      points: [[t, r.data.resources.search.remaining]],
    },
    {
      host: 'github.com',
      tags: [...tags, 'resource:search'],
      metric: `github.actions.api_rate_limit.limit`,
      type: 'gauge',
      points: [[t, r.data.resources.search.limit]],
    },
  ]

  if (r.data.resources.graphql) {
    series.push(
      {
        host: 'github.com',
        tags: [...tags, 'resource:graphql'],
        metric: 'github.actions.api_rate_limit.remaining',
        type: 'gauge',
        points: [[t, r.data.resources.graphql.remaining]],
      },
      {
        host: 'github.com',
        tags: [...tags, 'resource:graphql'],
        metric: 'github.actions.api_rate_limit.limit',
        type: 'gauge',
        points: [[t, r.data.resources.graphql.limit]],
      }
    )
  }

  return series
}

const unixTime = (s: string | undefined): number | undefined => {
  if (s === undefined) {
    return
  }
  const t = Date.parse(s)
  if (isNaN(t)) {
    return
  }
  return Math.floor(t / 1000)
}
