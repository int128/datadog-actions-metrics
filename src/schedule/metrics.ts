import { v1 } from '@datadog/datadog-api-client'
import { GitHubContext, ListWorkflowRunsForRepoRateLimitResponse } from '../types.js'

export const computeScheduleMetrics = (
  context: GitHubContext,
  queuedWorkflowRuns: ListWorkflowRunsForRepoRateLimitResponse,
  now: Date
): v1.Series[] => {
  const tags = [`repository_owner:${context.repo.owner}`, `repository_name:${context.repo.repo}`]
  const t = now.getTime() / 1000
  return [
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.schedule.queued_workflow_run.total',
      type: 'gauge',
      points: [[t, queuedWorkflowRuns.data.total_count]],
    },
  ]
}
