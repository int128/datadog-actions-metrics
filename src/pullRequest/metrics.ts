import { v1 } from '@datadog/datadog-api-client'
import { PullRequestClosedEvent, PullRequestEvent, PullRequestOpenedEvent } from '@octokit/webhooks-types'
import { PullRequestFirstCommit } from '../queries/getPullRequest.js'

const computeCommonTags = (e: PullRequestEvent): string[] => {
  const tags = [
    `repository_owner:${e.repository.owner.login}`,
    `repository_name:${e.repository.name}`,
    `sender:${e.sender.login}`,
    `sender_type:${e.sender.type}`,
    `user:${e.pull_request.user.login}`,
    `pull_request_number:${e.number}`,
    `draft:${String(e.pull_request.draft)}`,
    `base_ref:${e.pull_request.base.ref}`,
    `head_ref:${e.pull_request.head.ref}`,
  ]
  return tags
}

export const computePullRequestOpenedMetrics = (e: PullRequestOpenedEvent): v1.Series[] => {
  const tags = computeCommonTags(e)
  const t = unixTime(e.pull_request.created_at)
  return [
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_opened.total',
      type: 'count',
      points: [[t, 1]],
    },
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_opened.commits',
      type: 'count',
      points: [[t, e.pull_request.commits]],
    },
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_opened.changed_files',
      type: 'count',
      points: [[t, e.pull_request.changed_files]],
    },
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_opened.additions',
      type: 'count',
      points: [[t, e.pull_request.additions]],
    },
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_opened.deletions',
      type: 'count',
      points: [[t, e.pull_request.deletions]],
    },
  ]
}

type ClosedMetricsOptions = {
  sendPullRequestLabels: boolean
}

export const computePullRequestClosedMetrics = (
  e: PullRequestClosedEvent,
  pullRequestFirstCommit: PullRequestFirstCommit | undefined,
  options: ClosedMetricsOptions,
): v1.Series[] => {
  const tags = computeCommonTags(e)
  tags.push(`merged:${String(e.pull_request.merged)}`)
  const t = unixTime(e.pull_request.closed_at)
  const series: v1.Series[] = [
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_closed.total',
      type: 'count',
      points: [[t, 1]],
    },
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_closed.since_opened_seconds',
      type: 'gauge',
      points: [[t, t - unixTime(e.pull_request.created_at)]],
    },
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_closed.commits',
      type: 'count',
      points: [[t, e.pull_request.commits]],
    },
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_closed.changed_files',
      type: 'count',
      points: [[t, e.pull_request.changed_files]],
    },
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_closed.additions',
      type: 'count',
      points: [[t, e.pull_request.additions]],
    },
    {
      host: 'github.com',
      tags,
      metric: 'github.actions.pull_request_closed.deletions',
      type: 'count',
      points: [[t, e.pull_request.deletions]],
    },
  ]

  if (pullRequestFirstCommit !== undefined) {
    series.push(
      {
        host: 'github.com',
        tags,
        metric: 'github.actions.pull_request_closed.since_first_authored_seconds',
        type: 'gauge',
        points: [[t, t - unixTime(pullRequestFirstCommit.authoredDate)]],
      },
      {
        host: 'github.com',
        tags,
        metric: 'github.actions.pull_request_closed.since_first_committed_seconds',
        type: 'gauge',
        points: [[t, t - unixTime(pullRequestFirstCommit.committedDate)]],
      },
    )
  }

  // TODO: investigate how to analyze multi-value tag in Datadog
  //
  // When it sends multiple values, Datadog will show the tag as combination of values.
  // For example, it sends `label:x` and `label:y`, Datadog will show it as `label:x,y`.
  // Don't send each metric for a tag value, because sum of count metric would be wrong value.
  for (const requested_team of e.pull_request.requested_teams) {
    tags.push(`requested_team:${requested_team.name}`)
  }
  if (options.sendPullRequestLabels) {
    for (const label of e.pull_request.labels) {
      tags.push(`label:${label.name}`)
    }
  }

  return series
}

const unixTime = (s: string): number => Date.parse(s) / 1000
