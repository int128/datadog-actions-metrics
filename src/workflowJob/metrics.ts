import { v1 } from '@datadog/datadog-api-client'
import { WorkflowJobCompletedEvent } from '@octokit/webhooks-types'
import { WorkflowRun } from './queries.js'

type JobMetricsOptions = {
  preferDistributionJobMetrics: boolean
}

export type WorkflowJobMetrics = {
  series: v1.Series[]
  distributionPointsSeries: v1.DistributionPointsSeries[]
}

const joinRunsOn = (labels: string[]): string => labels.sort().join(',')

const getCommonMetricsTags = (e: WorkflowJobCompletedEvent, workflowRun: WorkflowRun): string[] => [
  `repository_owner:${e.repository.owner.login}`,
  `repository_name:${e.repository.name}`,
  `workflow_id:${workflowRun.id}`,
  `workflow_name:${workflowRun.name}`,
  `run_attempt:${workflowRun.run_attempt}`,
  `event:${workflowRun.event}`,
  `sender:${e.sender.login}`,
  `sender_type:${e.sender.type}`,
  `branch:${workflowRun.head_branch}`,
  `default_branch:${(workflowRun.head_branch === e.repository.default_branch).toString()}`,
  ...workflowRun.pull_requests.map((pull) => `pull_request_number:${pull.number}`),
  `job_id:${String(e.workflow_job.id)}`,
  `job_name:${e.workflow_job.name}`,
  `conclusion:${e.workflow_job.conclusion}`,
  `status:${e.workflow_job.status}`,
  `runs_on:${joinRunsOn(e.workflow_job.labels)}`,
]

const getCommonDistributionPointsTags = (e: WorkflowJobCompletedEvent, workflowRun: WorkflowRun): string[] => [
  `repository_owner:${e.repository.owner.login}`,
  `repository_name:${e.repository.name}`,
  `workflow_name:${workflowRun.name}`,
  `run_attempt:${workflowRun.run_attempt}`,
  `event:${workflowRun.event}`,
  `sender:${e.sender.login}`,
  `sender_type:${e.sender.type}`,
  `branch:${workflowRun.head_branch}`,
  `default_branch:${(workflowRun.head_branch === e.repository.default_branch).toString()}`,
  ...workflowRun.pull_requests.map((pull) => `pull_request_number:${pull.number}`),
  `job_name:${e.workflow_job.name}`,
  `conclusion:${e.workflow_job.conclusion}`,
  `status:${e.workflow_job.status}`,
  `runs_on:${joinRunsOn(e.workflow_job.labels)}`,
]

export const computeWorkflowJobMetrics = (
  e: WorkflowJobCompletedEvent,
  workflowRun: WorkflowRun,
  opts: JobMetricsOptions,
): WorkflowJobMetrics => {
  const series: v1.Series[] = []
  const distributionPointsSeries: v1.DistributionPointsSeries[] = []

  const tags = getCommonMetricsTags(e, workflowRun)
  const distributionPointsTags = getCommonDistributionPointsTags(e, workflowRun)

  const workflowRunStartedAt = unixTime(workflowRun.run_started_at)
  const jobStartedAt = e.workflow_job.started_at ? unixTime(e.workflow_job.started_at) : null
  const jobCompletedAt = unixTime(e.workflow_job.completed_at!)

  // Job total count
  series.push({
    host: 'github.com',
    tags,
    metric: 'github.actions.job.total',
    type: 'count',
    points: [[jobCompletedAt, 1]],
  })

  // Job conclusion count
  series.push({
    host: 'github.com',
    tags,
    metric: `github.actions.job.conclusion.${e.workflow_job.conclusion}_total`,
    type: 'count',
    points: [[jobCompletedAt, 1]],
  })

  // Job duration (if we have start time)
  if (jobStartedAt) {
    const duration = jobCompletedAt - jobStartedAt

    if (opts.preferDistributionJobMetrics) {
      distributionPointsSeries.push({
        host: 'github.com',
        tags: distributionPointsTags,
        metric: 'github.actions.job.duration_second.distribution',
        points: [[jobCompletedAt, [duration]]],
      })
    } else {
      series.push({
        host: 'github.com',
        tags,
        metric: 'github.actions.job.duration_second',
        type: 'gauge',
        points: [[jobCompletedAt, duration]],
      })
    }

    // Time from workflow start to job start
    const sinceWorkflowStart = jobStartedAt - workflowRunStartedAt

    if (opts.preferDistributionJobMetrics) {
      distributionPointsSeries.push({
        host: 'github.com',
        tags: distributionPointsTags,
        metric: 'github.actions.job.start_time_from_workflow_start_second.distribution',
        points: [[jobCompletedAt, [sinceWorkflowStart]]],
      })
    } else {
      series.push({
        host: 'github.com',
        tags,
        metric: 'github.actions.job.start_time_from_workflow_start_second',
        type: 'gauge',
        points: [[jobCompletedAt, sinceWorkflowStart]],
      })
    }
  }

  // Note: WorkflowJobCompletedEvent doesn't include step-level information
  // For step metrics, we would need to use the existing workflow_run event
  // or make additional API calls to get step details

  return { series, distributionPointsSeries }
}

const unixTime = (s: string): number => Date.parse(s) / 1000
