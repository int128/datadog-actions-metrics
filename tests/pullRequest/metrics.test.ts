import { computePullRequestClosedMetrics, computePullRequestOpenedMetrics } from '../../src/pullRequest/metrics.js'
import { examplePullRequestClosedEvent } from '../fixtures.js'
import { examplePullRequestOpenedEvent } from '../fixtures.js'
import { examplePullRequestFirstCommit } from './fixtures/getPullRequest.js'

test('computePullRequestOpenedMetrics', () => {
  const series = computePullRequestOpenedMetrics(examplePullRequestOpenedEvent)
  expect(series).toMatchSnapshot()
})

test('computePullRequestClosedMetrics', () => {
  const series = computePullRequestClosedMetrics(examplePullRequestClosedEvent, undefined, {
    sendPullRequestLabels: true,
  })
  expect(series).toMatchSnapshot()
})

test('computePullRequestClosedMetricsWithQuery', () => {
  const series = computePullRequestClosedMetrics(examplePullRequestClosedEvent, examplePullRequestFirstCommit, {
    sendPullRequestLabels: true,
  })
  expect(series).toMatchSnapshot()
})
