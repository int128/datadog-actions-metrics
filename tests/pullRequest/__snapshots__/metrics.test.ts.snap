// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computePullRequestClosedMetrics 1`] = `
[
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.total",
    "points": [
      [
        1557933678,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.since_opened_seconds",
    "points": [
      [
        1557933678,
        45,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "gauge",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.commits",
    "points": [
      [
        1557933678,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.changed_files",
    "points": [
      [
        1557933678,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.additions",
    "points": [
      [
        1557933678,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.deletions",
    "points": [
      [
        1557933678,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "count",
  },
]
`;

exports[`computePullRequestClosedMetricsWithQuery 1`] = `
[
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.total",
    "points": [
      [
        1557933678,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.since_opened_seconds",
    "points": [
      [
        1557933678,
        45,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "gauge",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.commits",
    "points": [
      [
        1557933678,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.changed_files",
    "points": [
      [
        1557933678,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.additions",
    "points": [
      [
        1557933678,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.deletions",
    "points": [
      [
        1557933678,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.since_first_authored_seconds",
    "points": [
      [
        1557933678,
        1267,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "gauge",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_closed.since_first_committed_seconds",
    "points": [
      [
        1557933678,
        596,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
      "merged:false",
      "label:bug",
    ],
    "type": "gauge",
  },
]
`;

exports[`computePullRequestOpenedMetrics 1`] = `
[
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_opened.total",
    "points": [
      [
        1557933633,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_opened.commits",
    "points": [
      [
        1557933633,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_opened.changed_files",
    "points": [
      [
        1557933633,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_opened.additions",
    "points": [
      [
        1557933633,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
    ],
    "type": "count",
  },
  {
    "host": "github.com",
    "metric": "github.actions.pull_request_opened.deletions",
    "points": [
      [
        1557933633,
        1,
      ],
    ],
    "tags": [
      "repository_owner:Codertocat",
      "repository_name:Hello-World",
      "sender:Codertocat",
      "sender_type:User",
      "user:Codertocat",
      "pull_request_number:2",
      "draft:false",
      "base_ref:master",
      "head_ref:changes",
    ],
    "type": "count",
  },
]
`;
